Add IsTrimmable property

sbomer opened the pull request:
This property will add IsTrimmable assembly metadata and enable the trim
analyzer. This doesn't set feature switches since libraries should not
in general be built with the assumption that feature settings have known
values.

Also adds a test that PublishTrimmed has the desired effect during build,
and doesn't block building without a RID. PublishTrimmed sets feature
switches and enables trim analysis but doesn't inject the assembly
attribute.

Fixes https://github.com/dotnet/sdk/issues/17201

sbomer committed:
Add IsTrimmable property

This property will add IsTrimmable assembly metadata and enable the trim
analyzer. This doesn't set feature switches since libraries should not
in general be built with the assumption that feature settings have known
values.

Also adds a test that PublishTrimmed has the desired effect during build,
and doesn't block building without a RID. PublishTrimmed sets feature
switches and enables trim analysis but doesn't inject the assembly
attribute.

This cleans up some of the related tests that don't need to set
SelfContained when publishing an exe project.

sbomer committed:
Clean up assertions

vitek-karas commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -1251,6 +1251,65 @@ class Program
+
+            var outputDirectory = buildCommand.GetOutputDirectory(targetFramework).FullName;
+            var assemblyPath = Path.Combine(outputDirectory, $"{projectName}.dll");
+            var runtimeConfigPath = Path.Combine(outputDirectory, $"{projectName}.runtimeconfig.json");
```
This will build an "app" test project, right?
In that case we should add another test case which builds "classlib" - as that is the primary target for this feature. (We won't be able to rely on runtimeconfig though).

sbomer replied:
I added a classlib test that just doesn't check the feature settings (since they don't affect libraries)

sbomer committed:
Add test for library project

sbomer committed:
Fix test failure

Avoid conflicts by ensuring that the parameterized test produces a
different test project on disk for each set of inputs. Also ensure the
other illink tests do this consistently in case we add more TFMs.

vitek-karas reviewed (approved):
LGTM
